lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,518 @@
+"""
+
+import logging
+from dataclasses import dataclass
+from typing import Any, Dict, List
+
+import ops
+from interface_tester import DataBagSchema
+from ops import Relation, RelationDataContent, SecretNotFoundError, model  # type: ignore
+from pydantic import BaseModel, Field, ValidationError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "c33e0a12506444e2b644ac2893ac9394"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 3
+
+
+class LogAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend a prefix to all log lines."""
+
+    prefix = "vault_autounseal"
+
+    def process(self, msg, kwargs):
+        """Decides the format for the prepended text."""
+        return f"[{self.prefix}] {msg}", kwargs
+
+
+logger = LogAdapter(logging.getLogger(__name__), {})
+
+
+class VaultAutounsealProviderSchema(BaseModel):
+    """Provider side of the vault-autounseal relation interface."""
+
+    address: str = Field(description="The address of the Vault server to connect to.")
+    mount_path: str = Field(
+        description="The path to the transit engine mount point where the key is stored."
+    )
+    key_name: str = Field(description="The name of the transit key to use for autounseal.")
+    credentials_secret_id: str = Field(
+        description=(
+            "The secret id of the Juju secret which stores the credentials for authenticating with the Vault server."
+        )
+    )
+    ca_certificate: str = Field(
+        description="The CA certificate to use when validating the Vault server's certificate."
+    )
+
+
+class ProviderSchema(DataBagSchema):
+    """The schema for the provider side of this interface."""
+
+    app: VaultAutounsealProviderSchema  # type: ignore
+
+
+class VaultAutounsealDetailsReadyEvent(ops.EventBase):
+    """Event emitted on the requirer when Vault autounseal details are ready in the databag."""
+
+    def __init__(
+        self, handle: ops.Handle, address, mount_path, key_name, role_id, secret_id, ca_certificate
+    ):
+        super().__init__(handle)
+        self.address = address
+        self.mount_path = mount_path
+        self.key_name = key_name
+        self.role_id = role_id
+        self.secret_id = secret_id
+        self.ca_certificate = ca_certificate
+
+    def snapshot(self) -> Dict[str, Any]:
+        """Return snapshot data that should be persisted."""
+        return dict(
+            super().snapshot(),
+            address=self.address,
+            mount_path=self.mount_path,
+            key_name=self.key_name,
+            role_id=self.role_id,
+            secret_id=self.secret_id,
+            ca_certificate=self.ca_certificate,
+        )
+
+    def restore(self, snapshot: Dict[str, Any]) -> None:
```

Comment:
Have you considered documenting the args?

Reply:
Added it in init.

Reply:
I don't see this specific change

Reply:
They are documented in the `init` function from line 128